dcicutils/misc_utils.py

```diff
@@ -368,112 +369,108 @@ def utc_today_str():
-#                 action_message = "Continuing anyway because lockout is disabled."
-#                 self.log.warning("%s %s" % (shared_message, action_message))
-#         self.update_timestamp()
-#
-#     def update_timestamp(self):
-#         """
-#         Explicitly sets the reference time point for computation of our lockout.
-#         This is called implicitly by .wait_if_needed(), and for some situations that may be sufficient.
-#         """
-#         self._timestamp = datetime.datetime.now()
+class LockoutManager:
+    """
+    This class is used as a guard of a critical operation that can only be called within a certain frequency.
+    e.g.,
+
+        class Foo:
+            def __init__(self):
+                # 60 seconds required between calls, with a 1 second margin of error (overhead, clocks varying, etc)
+                self.lockout_manager = LockoutManager(action="foo", lockout_seconds=1, safety_seconds=1)
+            def foo():
+                self.lockout_manager.wait_if_needed()
+                do_guarded_action()
+
+        f = Foo()     # make a Foo
+        v1 = f.foo()  # will immediately get a value
+        time.sleep(58)
+        v2 = f.foo()  # will wait about 2 seconds, then get a value
+        v3 = f.foo()  # will wait about 60 seconds, then get a value
+
+    Conceptually this is a special case of RateManager for n=1, though in practice it arose differently and
+    the supplementary methods (which we happen to use mostly for testing) differ because the n=1 case is simpler
+    and admits more questions. So, for now at least, this is not a subclass of RateManager but a separate
+    implementation.
+    """
+
+    EARLIEST_TIMESTAMP = datetime.datetime(datetime.MINYEAR, 1, 1)  # maybe useful for testing
+
+    def __init__(self, *, lockout_seconds, safety_seconds=0, action="metered action", enabled=True, log=None):
+        """
+        Creates a LockoutManager that cooperates in assuring a guarded operation is only happens at a certain rate.
+
+        The rate is once person lockout_seconds. This is a special case of RateManager and might get phased out
+        as redundant, but has slightly different operations available for testing.
+
+        Args:
+
+        lockout_seconds int: A theoretical number of seconds allowed between calls to the guarded operation.
+        safety_seconds int: An amount added to interval_seconds to accommodate real world coordination fuzziness.
+        action str: A noun or noun phrase describing the action being guarded.
+        enabled bool: A boolean controlling whether this facility is enabled. If False, waiting is disabled.
+        log object: A logger object (supporting operations like .debug, .info, .warning, and .error).
+        """
+
+        # This makes it easy to turn off the feature
+        self.lockout_enabled = enabled
+        self.lockout_seconds = lockout_seconds
+        self.safety_seconds = safety_seconds
+        self.action = action
```

Comment:
Would you set action to the name of the operation being guarded? Is this just for debugging?

Reply:
It's for error messages. It has no computational effect. (You might also have several of these and so you need a visual way to know which one you're talking to.)

Reply:
I think the doc string in the `__init__` method is pretty clear: "action str: A noun or noun phrase describing the action being guarded."